Packet++/src/Asn1Codec.cpp

```diff
@@ -0,0 +1,569 @@
+#define LOG_MODULE PacketLogModuleAsn1Codec
+
+#include "Asn1Codec.h"
+#include "GeneralUtils.h"
+#include "EndianPortable.h"
+#include <iostream>
+#include <cmath>
+#include <cstring>
+
+namespace pcpp {
+	std::unique_ptr<Asn1Record> Asn1Record::decode(const uint8_t* data, size_t dataLen, bool lazy)
+	{
+		auto record = decodeInternal(data, dataLen ,lazy);
+		return std::unique_ptr<Asn1Record>(record);
+	}
+
+	uint8_t Asn1Record::encodeTag()
+	{
+		uint8_t tagByte;
+
+		switch (m_TagClass)
+		{
+			case Asn1TagClass::Private:
+			{
+				tagByte = 0xc0;
+				break;
+			}
+			case Asn1TagClass::ContextSpecific:
+			{
+				tagByte = 0x80;
+				break;
+			}
+			case Asn1TagClass::Application:
+			{
+				tagByte = 0x40;
+				break;
+			}
+			default:
+			{
+				tagByte = 0;
+				break;
+			}
+		}
+
+		if (m_IsConstructed)
+		{
+			tagByte |= 0x20;
+		}
+
+		auto tagType = m_TagType & 0x1f;
+		tagByte |= tagType;
+
+		return tagByte;
+	}
+
+	std::vector<uint8_t> Asn1Record::encodeLength() const
+	{
+		std::vector<uint8_t> result;
+
+		if (m_ValueLength < 128)
+		{
+			result.push_back(static_cast<uint8_t>(m_ValueLength));
+			return result;
+		}
+
+		// Assuming the size is always 4 bytes
+		uint8_t firstByte = 0x80 | sizeof(uint32_t);
+		result.push_back(firstByte);
+
+		result.push_back((m_ValueLength >> 24) & 0xff);
+		result.push_back((m_ValueLength >> 16) & 0xff);
+		result.push_back((m_ValueLength >> 8) & 0xff);
+		result.push_back(m_ValueLength & 0xff);
+
+		return result;
+	}
+
+	std::vector<uint8_t> Asn1Record::encode()
+	{
+		std::vector<uint8_t> result;
+
+		result.push_back(encodeTag());
+
+		auto lengthBytes = encodeLength();
+		result.insert(result.end(), lengthBytes.begin(), lengthBytes.end());
+
+		auto encodedValue = encodeValue();
+		result.insert(result.end(), encodedValue.begin(), encodedValue.end());
+
+		return result;
+	}
+
+	Asn1Record* Asn1Record::decodeInternal(const uint8_t* data, size_t dataLen, bool lazy)
+	{
+		int tagLen;
+		auto decodedRecord = decodeTagAndCreateRecord(data, dataLen, tagLen);
+
+		int lengthLen;
+		try
+		{
+			lengthLen = decodedRecord->decodeLength(data + tagLen, dataLen - tagLen);
+		}
+		catch (...)
+		{
+			delete decodedRecord;
```

Comment:
Since I haven't checked the entire code I might miss a point but maybe `unique_ptr` can be used for 
`decodedRecord` and for return type of `decodeTagAndCreateRecord`?

Reply:
`decodeInternal()` returns a pointer that we can push into a `PointerVector` in case of a `Asn1ConstructedRecord`: https://github.com/seladb/PcapPlusPlus/blob/3dea318a164da4f3f723c917df70192ba73d7bd4/Packet%2B%2B/src/Asn1Codec.cpp#L390